elastalert/test_rule.py

```diff
@@ -1,86 +1,285 @@
         es_client = Elasticsearch(host=conf['es_host'], port=conf['es_port'])
-        for days in args.days:
-            start_time = ts_now() - datetime.timedelta(days=days)
-            end_time = ts_now()
-            ts = conf.get('timestamp_field', '@timestamp')
-            query = ElastAlerter.get_query(conf['filter'], starttime=start_time, endtime=end_time, timestamp_field=ts)
-            index = ElastAlerter.get_index(conf, start_time, end_time)
+        start_time = ts_now() - datetime.timedelta(days=args.days)
+        end_time = ts_now()
+        ts = conf.get('timestamp_field', '@timestamp')
+        query = ElastAlerter.get_query(conf['filter'], starttime=start_time, endtime=end_time, timestamp_field=ts)
+        index = ElastAlerter.get_index(conf, start_time, end_time)
+
+        # Get one document for schema
+        try:
+            res = es_client.search(index, size=1, body=query, ignore_unavailable=True)
+        except Exception as e:
+            logging.info("Error running your filter:")
+            logging.info(repr(e)[:2048])
+            return None
+        num_hits = len(res['hits']['hits'])
+        if not num_hits:
+            return []
+
+        terms = res['hits']['hits'][0]['_source']
+        doc_type = res['hits']['hits'][0]['_type']
+
+        # Get a count of all docs
+        count_query = ElastAlerter.get_query(conf['filter'], starttime=start_time, endtime=end_time, timestamp_field=ts, sort=False)
+        count_query = {'query': {'filtered': count_query}}
+        try:
+            res = es_client.count(index, doc_type=doc_type, body=count_query, ignore_unavailable=True)
+        except Exception as e:
+            logging.info("Error querying Elasticsearch:")
+            logging.info(repr(e)[:2048])
+            return None
+
+        num_hits = res['count']
+        logging.info("Got %s hits from the last %s day%s" % (num_hits, args.days, 's' if args.days > 1 else ''))
+        print_terms(terms, '')
+
+        # Check for missing keys
+        pk = conf.get('primary_key')
+        ck = conf.get('compare_key')
+        if pk and not lookup_es_key(terms, pk):
+            logging.info("Warning: primary key %s is either missing or null!")
+        if ck and not lookup_es_key(terms, ck):
+            logging.info("Warning: compare key %s is either missing or null!")
+
+        include = conf.get('include')
+        if include:
+            for term in include:
+                if not lookup_es_key(terms, term) and '*' not in term:
+                    logging.info("Included term %s may be missing or null" % (term))
+
+        for term in conf.get('top_count_keys', []):
+            # If the index starts with 'logstash', fields with .raw will be available but won't in _source
+            if term not in terms and not (term.endswith('.raw') and term[:-4] in terms and index.startswith('logstash')):
+                logging.info("top_count_key %s may be missing" % (term))
```

Comment:
I wish we had a schema for this instead of treating it like a special snowflake. Perhaps in a future change?